cwms-data-api/src/main/java/cwms/cda/data/dao/AuthDao.java

```diff
@@ -351,14 +353,13 @@ public void prepareGuestContext(Context ctx) {
         }
     }
 
-    private static String getFailMessage(@NotNull Context ctx,
+    private static String[] getMissingRoles(@NotNull Context ctx,
                                         @NotNull Set<RouteRole> requiredRoles,
                                         @NotNull DataApiPrincipal p) {
         Set<RouteRole> specifiedRoles = p.getRoles();
         Set<RouteRole> missing = new LinkedHashSet<>(requiredRoles);
         missing.removeAll(specifiedRoles);
-
-        return "Request for: " + ctx.req.getRequestURI() + " denied. Missing roles: " + missing;
+        return missing.stream().map(r -> r.toString()).collect(Collectors.toList()).toArray(new String[0]);
```

Comment:
Seems silly to return an array instead of a List when you're already creating a List. `stream().collect(join(","))` will do the same thing when processing the message.

Reply:
and in the interest of security, the list can be immutable while the array cannot be